File: go/src/AccessToken/AccessToken.go
package AccessToken

import (
	"fmt"
	"time"
	"math/rand"
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"encoding/hex"
	"io"
	"sort"
	"hash/crc32"
)

type Privileges uint16
const (
	KJoinChannel = 1
    KPublishAudioStream = 2
    KPublishVideoStream = 3
    KPublishDataStream = 4

	KPublishAudiocdn = 5
    KPublishVideoCdn = 6
    KRequestPublishAudioStream = 7
    KRequestPublishVideoStream = 8
    KRequestPublishDataStream = 9
    KInvitePublishAudioStream = 10
    KInvitePublishVideoStream = 11
	KInvitePublishDataStream = 12
	
	KAdministrateChannel = 101
)

type AccessToken struct {
	AppID          string
	AppCertificate string
	ChannelName string
	UidStr  string
	Ts  uint32
	Salt uint32
	Message map[uint16]uint32
}

func random(min int, max int) int {
	rand.Seed(time.Now().UnixNano())
    return rand.Intn(max-min) + min
}

func CreateAccessToken(appID, appCertificate, channelName string, uid uint32) AccessToken {
	uidStr := fmt.Sprintf("%d", uid)
	ts := uint32(time.Now().Unix()) + 24 * 3600
	salt := uint32(random(1, 99999999))
	message := make(map[uint16]uint32)
    return AccessToken{appID, appCertificate, channelName, uidStr, ts, salt, message}
}

func (token AccessToken) AddPrivilege(key Privileges, secondsFromNow uint32) {
	pri := uint16(key)
	token.Message[pri] = uint32(time.Now().Unix()) + secondsFromNow
}

func (token AccessToken) Build() (string, error) {
	ret := ""
	version := "006"

	buf_m := new(bytes.Buffer)
	if err := packUint32(buf_m, token.Salt); err != nil {
		return ret, err
	}
	if err := packUint32(buf_m, token.Ts); err != nil {
		return ret, err
	}
	if err := packMapUint32(buf_m, token.Message); err != nil {
		return ret, err
	}
	bytes_m := buf_m.Bytes()

	buf_val := new(bytes.Buffer)
	val := token.AppID + token.ChannelName + token.UidStr
	buf_val.Write([]byte(val))
	buf_val.Write(bytes_m)
	bytes_val := buf_val.Bytes()

	buf_sig := hmac.New(sha256.New, []byte(token.AppCertificate))
	buf_sig.Write(bytes_val)
	bytes_sig := buf_sig.Sum(nil)

	crc32q := crc32.MakeTable(0xedb88320)
	crc_channel_name := crc32.Checksum([]byte(token.ChannelName), crc32q)
	crc_uid := crc32.Checksum([]byte(token.UidStr), crc32q)

	buf_content := new(bytes.Buffer)
	if err := packString(buf_content, string(bytes_sig[:])); err != nil {
		return ret, err
	}
	if err := packUint32(buf_content, crc_channel_name); err != nil {
		return ret, err
	}
	if err := packUint32(buf_content, crc_uid); err != nil {
		return ret, err
	}
	if err := packString(buf_content, string(bytes_m[:])); err != nil {
		return ret, err
	}
	bytes_content := buf_content.Bytes()

	ret = version + token.AppID + base64.StdEncoding.EncodeToString(bytes_content)
	return ret, nil
}


func packUint16(w io.Writer, n uint16) error {
	return binary.Write(w, binary.LittleEndian, n)
}

func packUint32(w io.Writer, n uint32) error {
	return binary.Write(w, binary.LittleEndian, n)
}

func packString(w io.Writer, s string) error {
	err := packUint16(w, uint16(len(s)))
	if err != nil {
		return err
	}
	_, err = w.Write([]byte(s))
	return err
}

func packHexString(w io.Writer, s string) error {
	b, err := hex.DecodeString(s)
	if err != nil {
		return err
	}
	return packString(w, string(b))
}

func packExtra(w io.Writer, extra map[uint16]string) error {
	keys := []int{}
	if err := packUint16(w, uint16(len(extra))); err != nil {
		return err
	}
	for k := range extra {
		keys = append(keys, int(k))
	}
	//should sorted keys
	sort.Ints(keys)

	for _, k := range keys {
		v := extra[uint16(k)]
		if err := packUint16(w, uint16(k)); err != nil {
			return err
		}
		if err := packString(w, v); err != nil {
			return err
		}
	}
	return nil
}

func packMapUint32(w io.Writer, extra map[uint16]uint32) error {
	keys := []int{}
	if err := packUint16(w, uint16(len(extra))); err != nil {
		return err
	}
	for k := range extra {
		keys = append(keys, int(k))
	}
	//should sorted keys
	sort.Ints(keys)

	for _, k := range keys {
		v := extra[uint16(k)]
		if err := packUint16(w, uint16(k)); err != nil {
			return err
		}
		if err := packUint32(w, v); err != nil {
			return err
		}
	}
	return nil
}

File: php/src/AccessToken.php
<?php


$Privileges = array(
    "kJoinChannel" => 1,
    "kPublishAudioStream" => 2,
    "kPublishVideoStream" => 3,
    "kPublishDataStream" => 4,
    "kPublishAudiocdn" => 5,
    "kPublishVideoCdn" => 6,
    "kRequestPublishAudioStream" => 7,
    "kRequestPublishVideoStream" => 8,
    "kRequestPublishDataStream" => 9,
    "kInvitePublishAudioStream" => 10,
    "kInvitePublishVideoStream" => 11,
    "kInvitePublishDataStream" => 12,
    "kAdministrateChannel" => 101
);

class Message
{
    private $salt;
    private $ts;
    private $msgs;
    public function __construct($salt, $ts, $msgs)
    {
        $this->salt = $salt;
        $this->ts = $ts;
        $this->msgs = $msgs;
    }

    public function packContent()
    {
        $buffer = unpack("C*", pack("I", $this->salt));
        $buffer = array_merge($buffer, unpack("C*", pack("I", $this->ts)));
        $buffer = array_merge($buffer, unpack("C*", pack("S", sizeof($this->msgs))));
        foreach ($this->msgs as $key => $value) {
            $buffer = array_merge($buffer, unpack("C*", pack("S", $key)));
            $buffer = array_merge($buffer, unpack("C*", pack("I", $value)));
        }
        return $buffer;
    }
}

class AccessToken
{
    private $msgs;
    private $appID, $appCertificate, $channelName, $uid;
    public $ts, $salt;
    public function __construct($appID, $appCertificate, $channelName, $uid)
    {
        $this->appID = $appID;
        $this->appCertificate = $appCertificate;
        $this->channelName = $channelName;
        $this->uid = $uid;
        date_default_timezone_set("UTC");
        $date = new DateTime();
        $this->ts = $date->getTimestamp() + 24 * 3600;
        $this->salt = rand(0, 100000);
        $this->msgs = array();
    }

    public function setPriviledge($key, $seconds)
    {
        $this->msgs[$key] = $seconds;
        return $this;
    }

    public function removePriviledge($key, $seconds)
    {
        unset($this->msgs[$key]);
        return $this;
    }

    public function build()
    {
        $msg = (new Message($this->salt, $this->ts, $this->msgs))->packContent();
        $val = array_merge(unpack("C*", $this->appID), unpack("C*", $this->channelName), unpack("C*", $this->uid), $msg);
        $sig = hash_hmac('sha256', implode(array_map("chr", $val)), $this->appCertificate, true);

        $crc_channel_name = crc32($this->channelName) & 0xffffffff;
        $crc_uid = crc32($this->uid) & 0xffffffff;

        $content = array_merge(unpack("C*", packString($sig)), unpack("C*", pack("I", $crc_channel_name)), unpack("C*", pack("I", $crc_uid)), unpack("C*", pack("S", count($msg))), $msg);
        $version = "006";
        $ret = $version . $this->appID . base64_encode(implode(array_map("chr", $content)));
        return $ret;
    }
}

function packString($value)
{
    return pack("S", strlen($value)) . $value;
}
